### PR TITLE
Zhu 2023: 500 markers

### LINKED ISSUE #123: Zhu et al. 500 MH / RMNE
*Thank you for your suggestion, we are eager to include all published microhaplotype data in MicroHapDB! Please provide the information below to facilitate its integration into the database.*

## Paper

Zhu Q, Wang H, Cao Y, Huang Y, Wei Y, Hu Y, Dai X, Shan T, Wang Y, Zhang J (2023) Evaluation of large-scale highly polymorphic microhaplotypes in complex DNA mixtures analysis using RMNE method. *FSI: Genetics* 65:102874, [doi:10.1016/j.fsigen.2023.102874](https://doi.org/10.1016/j.fsigen.2023.102874).

## Data URL(s)

- https://doi.org/10.1016/j.fsigen.2023.102874
- [mmc2.csv](https://github.com/bioforensics/MicroHapDB/files/11313026/mmc2.csv)
- [mmc2.xlsx](https://github.com/bioforensics/MicroHapDB/files/11313027/mmc2.xlsx)

## Author and PI information

- Yunfeng Wang yfwang@scu.edu.cn
- Ji Zhang zhangj@scu.edu.cn

### PR DESCRIPTION
**Pull request by standage**
This PR integrates markers from the recently published Zhu 2023 study (#123). The 500 markers defined in this study break down as follows:

- 6 were excluded due to falling within regions that cannot be mapped directly from GRCh37 to GRCh38, either in part or in whole
- 40 overlap with markers defined in previous studies; 5 of these 40 are exact duplicates of previously published markers
- 454 occupy unique loci for which no marker has been previously described

The marker names used in the study did not follow the Kidd 2016 nomenclature. The corresponding author on the study has agreed to adopting the community's *de facto* standard nomenclature, and has asked that the markers use the lab label "SCUZJ" (referring to SiChuan University and Zhang Ji). Thus, `CHSCHR04_0684549` in the study becomes `mh04SCUZJ-0684549`.

Closes #123.

### CODE REVIEW

File: dbbuild/sources/zhu2023/table-s1.csv [hunk 1]
```diff
@@ -0,0 +1,502 @@
...
+CHSCHR10_0290121,340,5,0.866666667,0.923356009,5,8,72046476,72046512,72046609,72046798,72046816,CACAC:0.2,CACCC:0.32857142857142857,CGCCC:0.08571428571428572,TACCC:0.10952380952380952,TACAC:0.10952380952380952,CGCCT:0.14285714285714285,CGCAC:0.009523809523809525,CATAC:0.014285714285714285,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,
+CHSCHR21_0003240,341,5,0.761904762,0.920272109,8,13,15243139,15243242,15243263,15243339,15243353,15243433,15243447,15243480,CGCAGCCG:0.08095238095238096,CCCCGCCG:0.06190476190476191,CGCCGCCG:0.21904761904761905,CGGAGCTC:0.319047619047619,CGGAATTC:0.19047619047619047,TGGAGCTC:0.03333333333333333,CGGCATTC:0.01904761904761905,CGGCGCTC:0.04285714285714286,CGGAGCCC:0.009523809523809525,CCCAGCCG:0.009523809523809525,TGGCGCTC:0.004761904761904762,CGGAGCCG:0.004761904761904762,CGGAATCG:0.004761904761904762,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,
```
Comment (standage) [inline dbbuild/sources/zhu2023/table-s1.csv:5]: Authors did not attempt to associate SNPs with RSIDs, they provided only GRCh37(!) coordinates.

File: dbbuild/lib/marker.py [hunk 2]
```diff
@@ -36,7 +36,9 @@ class Marker:
     def from_csv(csvpath, index, source=None):
         table = pd.read_csv(csvpath)
         for n, row in table.iterrows():
-            rsids = row.VarRef.split(";")
+            rsids = []
+            if not pd.isna(row.VarRef):
+                rsids = row.VarRef.split(";")
```
Comment (standage) [inline dbbuild/lib/marker.py:41]: This piece of dbbuild code was failing with the new data. It successfully handled marker definitions where 1 or more RSIDs were missing, e.g. for rare variants lacking an RSID. The code was updated in this PR to correctly handle marker definitions providing 0 RSIDs.

File: microhapdb/cli/summarize.py [hunk 3]
```diff
@@ -11,16 +11,14 @@
...
-from textwrap import dedent
 
 
 def main(args):
     num_markers = len(microhapdb.markers)
     markers = microhapdb.Marker.objectify(microhapdb.markers)
     num_loci = len(set([m.locus for m in markers]))
     print("[microhaplotypes]")
-    print(f"  - {num_markers} marker defintions")
+    print(f"  - {num_markers} marker definitions")
```
Comment (standage) [inline microhapdb/cli/summarize.py:21]: Fixed typo

File: microhapdb/nomenclature.py [hunk 4]
```diff
@@ -109,7 +109,7 @@ def validate_lab(self):
         lab = self._raw[4:].split("-")[0]
         if not lab.isupper():
             self._validation_warnings.append(ValidationWarnings.LABCASE)
-        if len(lab) < 2 or len(lab) > 4:
+        if len(lab) < 2 or len(lab) > 6:
```
Comment (standage) [inline microhapdb/nomenclature.py:112]: The corresponding author requested the label "SCUZJ" for these markers. The Kidd 2016 nomenclature stipulates 2-4 characters for the lab ID: I think 4 is a reasonable limit, but I also think the author's request is reasonable and I don't think there's a reason for strict enforcement at the moment. I relaxed the hard limit to 6, while maintaining the rhetorical limit at 4.

File: microhapdb/tests/test_marker.py [hunk 5]
```diff
@@ -70,6 +71,7 @@ def test_assumptions():
         296,  # Yu2022G2
         85,  # Yu2022G3
         64,  # Yu2022G4
+        5,  # Zhu2023
```
Comment (standage) [inline microhapdb/tests/test_marker.py:74]: These are the 5 markers that are exact duplicates of previously published markers.

File: microhapdb/tests/test_nomenclature.py [hunk 6]
```diff
@@ -64,7 +64,7 @@ def test_hyphen():
 
 
 def test_lab():
-    id1 = Identifier("mh16DSSAGS-123")
+    id1 = Identifier("mh16DSSAGSRNM-123")
```
Comment (standage) [inline microhapdb/tests/test_nomenclature.py:67]: Updated this test to trigger the expected validation failure due to a long lab label.

**Comment by rnmitchell** [comment]
> Ae as a function of marker length for 494 microhaps described in the study.
> 
> ae_vs_length_shu2023
> 
> Twenty of the markers have Ae≥8 for the Ae value aggregated over all 26 1000 Genomes Project populations. (Two of these markers were previously published under different names: mh03SCUZJ-0030960 as mh03LV-01, and mh12SCUZJ-0259712 as mh12WL-005).
> 
> Name	Extent	Source	Ae
> 378	mh03SCUZJ-0000972	300	Zhu2023	8.882
> 387	mh03LV-01.v2	139	Zhu2023	11.734
> 403	mh03SCUZJ-0080612	231	Zhu2023	15.241
> 645	mh04SCUZJ-0550345	329	Zhu2023	8.048
> 810	mh06SCUZJ-0000232	251	Zhu2023	8.8
> 884	mh06SCUZJ-0610307	292	Zhu2023	8.062
> 1054	mh07SCUZJ-0502291	225	Zhu2023	9.083
> 1110	mh08SCUZJ-0190805	285	Zhu2023	8.024
> 1238	mh09SCUZJ-0251127	338	Zhu2023	10.271
> 1420	mh11SCUZJ-0062178	254	Zhu2023	9.186
> 1574	mh12WL-005.v2	184	Zhu2023	9.309
> 1725	mh14SCUZJ-0040957	195	Zhu2023	9.166
> 1740	mh14SCUZJ-0096560	196	Zhu2023	8.059
> 1852	mh15SCUZJ-0204763	240	Zhu2023	9.158
> 1863	mh15SCUZJ-0251800	133	Zhu2023	8.162
> 1905	mh16SCUZJ-0113007	211	Zhu2023	9.879
> 1935	mh16SCUZJ-0260785	291	Zhu2023	8.838
> 1976	mh17SCUZJ-0017256	324	Zhu2023	8.218
> 2207	mh20SCUZJ-0148996	91	Zhu2023	10.818
> 2246	mh21SCUZJ-0093029	174	Zhu2023	10.575

The reference in the plot says ```Zhu 2021```. I assume it's supposed to be ```Zhu 2023```?

**Comment by rnmitchell** [comment]
I'm running into the following error:
```
Exception ignored in: <function Faidx.__del__ at 0x1144fa820>
Traceback (most recent call last):
  File "/Users/rebecca.mitchell/mambaforge/envs/microhapdb/lib/python3.9/site-packages/pyfaidx/__init__.py", line 889, in __del__
    self.__exit__()
  File "/Users/rebecca.mitchell/mambaforge/envs/microhapdb/lib/python3.9/site-packages/pyfaidx/__init__.py", line 895, in __exit__
    self.file.close()
AttributeError: 'Faidx' object has no attribute 'file'
```

What version of ```pyfaidx``` do you have installed? I have ```0.7.2.1```.

**Comment by standage** [comment]
> What version of `pyfaidx` do you have installed? I have `0.7.2.1`.

Monday was the first time I've had a chance to work on microhaps since March, and in the meantime there's been an update to pyfaidx that breaks compatibility with MicroHapDB. It looks like we need to keep it at 0.7.1 for now. I'll create an issue thread to track that.

**Comment by standage** [comment]
> It looks like we need to keep it at 0.7.1 for now. I'll create an issue thread to track that.

Pinned in 36da308, tracked in #127.

**Comment by rnmitchell** [comment]
```
(microhapdb) FDT010LM-613658:~ rebecca.mitchell$ microhapdb frequency --marker mh14SCUZJ-0040957 | head
Marker	Population	Allele	Frequency	Source
mh14SCUZJ-0040957	1KGP	G|C|A|A|A|C|T|G|T|G	0.00038	Byrska-Bishop2022
mh14SCUZJ-0040957	1KGP	G|C|A|A|A|T|T|G|C|G	0.00019	Byrska-Bishop2022
mh14SCUZJ-0040957	1KGP	G|C|A|G|A|T|T|G|T|G	0.00019	Byrska-Bishop2022
mh14SCUZJ-0040957	1KGP	G|C|G|A|A|C|C|A|C|G	0.00019	Byrska-Bishop2022
mh14SCUZJ-0040957	1KGP	G|C|G|A|A|C|C|G|C|G	0.00115	Byrska-Bishop2022
mh14SCUZJ-0040957	1KGP	G|C|G|A|A|C|C|G|T|G	0.00058	Byrska-Bishop2022
mh14SCUZJ-0040957	1KGP	G|C|G|A|A|C|T|A|C|G	0.00019	Byrska-Bishop2022
mh14SCUZJ-0040957	1KGP	G|C|G|A|A|C|T|G|C|G	0.00192	Byrska-Bishop2022
mh14SCUZJ-0040957	1KGP	G|C|G|A|A|C|T|G|T|G	0.00019	Byrska-Bishop2022
```

Why is the source not ```Zhu2023```?

**Comment by standage** [comment]
> Why is the source not `Zhu2023`?

Good question. While the source of the *markers* is indeed Zhu 2023, The source of the *frequencies* is the NYGC 1KGP data.

**Review by rnmitchell — APPROVED** [review]

**Pull request merged**